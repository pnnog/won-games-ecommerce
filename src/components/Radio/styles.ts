import styled, { css } from 'styled-components';
import { RadioProps } from '.';

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

export const Input = styled.input`
  ${({ theme }) => css`
    border: 0.2rem solid ${theme.colors.primary};
    line-height: 1;
    outline: 0;
    position: relative;

    border-radius: 50%;
    appearance: none;
    cursor: pointer;
    background: transparent;

    height: 1.8rem;
    width: 1.8rem;

    :before {
      content: '';
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      width: 0.8rem;
      height: 0.8rem;
      background: ${theme.colors.primary};
      border-radius: 50%;
      opacity: 0;
      transition: opacity ${theme.transition.fast};
    }

    :checked {
      :before {
        opacity: 1;
      }
    }

    :focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

type LabelProps = Pick<RadioProps, 'labelColor'>;
export const Label = styled.label<LabelProps>`
  ${({ theme, labelColor }) => css`
    color: ${theme.colors[labelColor!]};
    margin-left: ${theme.spacings.xxsmall};
  `}
`;
