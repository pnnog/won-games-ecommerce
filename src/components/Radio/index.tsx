import { InputHTMLAttributes } from 'react';
import * as S from './styles';

type RadioValue = string | ReadonlyArray<string> | number | undefined;

export type RadioProps = {
  onCheck?: (value: RadioValue) => void;
  value?: RadioValue;
  label?: string;
  labelFor?: string;
  labelColor?: 'black' | 'white';
} & InputHTMLAttributes<HTMLInputElement>;

const Radio = ({
  onCheck,
  value,
  label,
  labelFor = '',
  labelColor = 'white',
  ...props
}: RadioProps) => {
  const handleOnChange = () => {
    onCheck!(value);
  };

  return (
    <S.Wrapper>
      <S.Input
        id={labelFor}
        type="radio"
        value={value}
        onChange={handleOnChange}
        {...props}
      />
      {!!label && (
        <S.Label labelColor={labelColor} htmlFor={labelFor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  );
};

export default Radio;
