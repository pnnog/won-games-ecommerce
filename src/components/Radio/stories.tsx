import { Meta, Story } from '@storybook/react';
import Radio, { RadioProps } from '.';

export default {
  title: 'Forms/Radio',
  component: Radio,
  args: {},

  argTypes: {
    onCheck: { action: 'checked' },
  },

  parameters: {
    viewport: {
      defaultViewport: 'mobile1',
    },

    layout: {
      default: 'fullscreen',
    },
  },
} as Meta;

export const Default: Story<RadioProps> = (args) => (
  <>
    <div style={{ padding: '2rem' }}>
      <Radio
        name="radio"
        label="First Radio"
        value="First Radio"
        labelFor="First Radio"
        {...args}
      />
    </div>

    <div style={{ padding: '2rem' }}>
      <Radio
        name="radio"
        label="Secondary Radio"
        value="Secondary Radio"
        labelFor="Secondary Radio"
        {...args}
      />
    </div>

    <div style={{ padding: '2rem' }}>
      <Radio
        name="radio"
        label="Terciary Radio"
        value="Terciary Radio"
        labelFor="Terciary Radio"
        {...args}
      />
    </div>
  </>
);
