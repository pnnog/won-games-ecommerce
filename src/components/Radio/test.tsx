import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Radio from '.';
import { renderWithTheme } from 'utils/tests/helpers';
import theme from 'styles/theme';

describe('<Radio />', () => {
  it('should render default Radio', () => {
    const { container } = renderWithTheme(
      <Radio label="label" labelFor="labelFor" value="label" />
    );

    expect(screen.getByRole('radio', { name: 'label' })).toBeInTheDocument();

    expect(screen.getByText('label')).toHaveStyle({
      color: theme.colors.white,
    });

    expect(screen.getByText('label')).toHaveAttribute('for', 'labelFor');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render Radio with black label', () => {
    renderWithTheme(
      <Radio labelFor="labelFor" label="label" labelColor="black" />
    );
    expect(screen.getByText('label')).toHaveStyle({
      color: theme.colors.black,
    });
  });

  it('should render radio callback when status is changed', async () => {
    //mock a função e passando para o componente
    const mockFn = jest.fn();
    renderWithTheme(
      <Radio
        labelFor="radio"
        label="radio"
        onCheck={mockFn}
        value="any value"
      />
    );
    //verifica chamada a função
    expect(mockFn).not.toBeCalled();

    //altera o status com click
    await waitFor(() => userEvent.click(screen.getByLabelText('radio')));

    //verifica chamada a função após o click
    expect(mockFn).toHaveBeenCalledTimes(1);

    //verifica se o argumento recebido pela função é igual o value
    expect(mockFn).toHaveBeenCalledWith('any value');
  });
});
