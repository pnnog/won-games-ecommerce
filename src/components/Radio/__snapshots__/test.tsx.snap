// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render default Radio 1`] = ` `;
